### DISCUSSION TITLE
Nuxt3 lifecycle hook after routing

### DISCUSSION THREAD

**Discussion by amos81** [nuxt/nuxt#16730] [Questions]
I'm porting an existing front-end into Nuxt, I started with 2 and am now trying 3. I have several components that I want to call a JS function after the DOM elements are loaded. In Nuxt 2 I just called the functions at the mounted lifecycle hook, it worked fine. In Nuxt 3 those functions are called whenever the page first loads or on manual refresh, but never after navigating to other pages. Is there another lifecycle hook that I could be using?

Here's an example in the component's script setup:
	onMounted( () => {
		inPageNavigation()
	})

**Comment by fabianbeier** [comment]
Hi! Just had the same Issue the other day! A gsap-function did not fire after route-change. 
I don't really have an explanation - only a work-around. 

```
definePageMeta({
  pageTransition: {
    mode: "default",
    appear: true,
  },
});
```
For me it started working after i used **definePageMeta** with mode "default". But please don't ask why 🥴

**Reply by fabianbeier** [reply]
A sandbox for demonstration. If you delete definePageMeta({...}), the animation wont run.
https://codesandbox.io/s/fervent-shadow-f7gtyh?file=/pages/second.vue

**Reply by shaheer-haider** [reply]
Thanks for your answer bro and don't worry I am not gonna ask why. I was using a transition element with an after-enter hook everywhere. Thanks to save me.

**Reply by shaheer-haider** [reply]
I tried it today but not working in my case.
Here's the code of my component, this one is working for the first time but not after the coming from another route to this one.

<template>
    <h1 id="animTitle">Not animating</h1>
</template>

<script setup>
import { onMounted } from "vue"
import gsap from "gsap"

definePageMeta({
  pageTransition: {
    mode: "default",
    appear: true,
  },
});

onMounted(() => {
    gsap.fromTo("#animTitle", {
        duration: 0,
        opacity: 0,
        rotation: 0,
        scale: 0,
        y: 100,
    }, {
        duration: 1.5,
        opacity: 1,
        rotation: 360,
        scale: 1,
        repeat: 0,
        y: 0,
        ease: "linear",
    })
})

</script>

Here's the working code with a bad approach



<template>
    <transition @after-enter="afterEnter">
        <h1 v-show="show" id="animTitle">Animating but with a bad approach, I think</h1>
        </transition>
</template>

<script setup>
import { ref } from "vue"
import gsap from "gsap"

const show = ref(false)
setTimeout(() => {
    show.value = true
}, 0)
const afterEnter = () => {
    gsap.fromTo("#animTitle", {
        duration: 0,
        opacity: 0,
        rotation: 0,
        scale: 0,
        y: 100,
    }, {
        duration: 1.5,
        opacity: 1,
        rotation: 360,
        scale: 1,
        repeat: 0,
        y: 0,
        ease: "linear",
    })
}
</script>

**Reply by amos81** [reply]
Thanks for your answer but that's not quite working for my issue. That only works on the page, not the components or the layout. Even putting it on every page the onMounted still doesn't trigger consistently, some of my functions work but not others.

Did you report this as a bug already?

**Reply by shaheer-haider** [reply]
No, I didn't but please if you do, send its link here in comments too.

**Reply by crunchwrap89** [reply]
I have the same issue.
Fabian's solution works for me, but it also worked if i removed these classes from tailwind instead:

```
.page-enter-active,
.page-leave-active {
	transition: all 0.3s ease-in-out;
}

.page-enter-from,
.page-leave-to {
	opacity: 0;
	transform: scale(0.99);
}
```

**Comment by RyonMB** [comment]
I have the same issue! But it's triggering mounted before the dom is actually loaded. So it causes to trigger code that should only run after the page is loaded. Like Stripe Elements

**Comment by amos81** [comment]
It looks like the latest update to Nuxt 3 has fixed this bug. Thanks.

**Comment by dylanmcgowan** [comment]
This is still an issue as of 3.0.0-rc.9 :(

GSAP animations are not animating on route changes only on hard refreshes.

And defining `pageTransition: { mode: default }` in nuxt config does not work. You must define it inside the page component, per page.

Any suggestions/fixes?

Thanks.

**Reply by bennyzen** [reply]
Had the same issue and solved it globally by adding this to my `nuxt.config.ts`:
```js
app: {
    ...
    // this fixes GSAP animations in onMounted hooks
    layoutTransition: false,
    pageTransition: false,
  },
  ```
  It's maybe not the smartest fix, but worked well for me. It fixed issues with other libs after route change as well.

**Reply by amos81** [reply]
You are right, it hasn't worked for several RCs now. The layout and page transitions set to false in nuxt.config.ts works as a solution.

Another hacky solution is just add a time delay to your onMounted callback:
`
onMounted( () => {
  setTimeout( () => {
    // do the thing
  }, 500)
}
`

I'm watching this issue for an actual fix to this bug: https://github.com/nuxt/framework/issues/3587
